From 12055d6232f84470406929189319dfb42f65a175 Mon Sep 17 00:00:00 2001
From: Vates Git Kernel Importer <gitimporter@vates.tech>
Date: Sat, 19 Sep 2015 23:58:42 +0200
Subject: [PATCH] sunrpc-force-disconnect-on-connection-timeout.patch

[Quentin: rebase from 4.19.19 to 4.19.325:
 - 6cedfaffb4ac ("sunrpc: Fix connect metrics") removed the local variable
   xprt which is used in this commit, re-introduce it.]

Signed-off-by: Quentin Casasnovas <quentin.casasnovas@vates.tech>
---
 net/sunrpc/xprt.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/net/sunrpc/xprt.c b/net/sunrpc/xprt.c
index 99c4a7274f60..9508db033c77 100644
--- a/net/sunrpc/xprt.c
+++ b/net/sunrpc/xprt.c
@@ -897,6 +897,8 @@ void xprt_connect(struct rpc_task *task)
 
 static void xprt_connect_status(struct rpc_task *task)
 {
+	struct rpc_xprt *xprt = task->tk_rqstp->rq_xprt;
+
 	switch (task->tk_status) {
 	case 0:
 		dprintk("RPC: %5u xprt_connect_status: connection established\n",
@@ -914,6 +916,16 @@ static void xprt_connect_status(struct rpc_task *task)
 	case -ETIMEDOUT:
 		dprintk("RPC: %5u xprt_connect_status: connect attempt timed "
 				"out\n", task->tk_pid);
+		/*
+		 * Timed out while waiting for the old connection to
+		 * be closed?
+		 *
+		 * Force a disconnect since its unlikely to close
+		 * gracefully and waiting for the socket close to
+		 * timeout may take a long time.
+		 */
+		if (test_bit(XPRT_CLOSING, &xprt->state))
+			xprt_force_disconnect(xprt);
 		break;
 	default:
 		dprintk("RPC: %5u xprt_connect_status: error %d connecting to "
-- 
2.43.0

